Coding/server/lib/User/User.cpp

```diff
@@ -0,0 +1,301 @@
+                {
+                    retCode = false;
+                }
+            }
+        }
+        else
+        {
+            retCode = false;
+        }
+    }
+    else
+    {
+        retCode = false;
+    }
+    return retCode;
+}
+
+bool User::putUser(const String& username, const String& password, const Permission* permissions, const uint8_t numberOfPermissions, const bool updateUser)
+{
+    String randomSalt;
+    String hashedPassword;
+    bool retCode = false;
+    bool deleted = true;
+
+    /* Fully re-create user if it needs to be updated */
+    if (true == updateUser)
+    {
+        deleted = deleteUser(username);
+    }
+
+    if ((true == deleted)
+        && (-1 == getUserIdx(username))
+        && (m_numberOfRegisteredUsers < MAX_REGISTERED_USERS)
+        && ((MIN_USERNAME_CHARS <= username.length()) && (username.length() <= MAX_USERNAME_CHARS))
+        && ((MIN_PASSWORD_CHARS <= password.length()) && (password.length() <= MAX_PASSWORD_CHARS)))
+    {
+        for (uint8_t userIdx = 0; userIdx < MAX_REGISTERED_USERS; userIdx++)
+        {
+            if (nullptr == m_registeredUsers[userIdx])
+            {
+                User* newUser = new User();
+                newUser->m_username = username;
+
+                /* Generate a unique and random salt for hash */
+                m_crypto.getRandomSalt(randomSalt);
+                newUser->m_passwordSalt = randomSalt;
+
+                /* Calculate the salted hash */
+                m_crypto.hashBlake2b(password, randomSalt, hashedPassword);
+                newUser->m_hashedPassword = hashedPassword;
+
+                newUser->m_numberOfPermissions = numberOfPermissions;
+
+                /* Deep copy permissions */
+                for (uint8_t permIdx = 0; permIdx < numberOfPermissions; permIdx++)
+                {
+                    newUser->m_permissions[permIdx] = permissions[permIdx];
+                }
+
+                m_registeredUsers[userIdx] = newUser;
+                m_numberOfRegisteredUsers++;
+                retCode = true;
+                break;
+            }
+        }
+    }
+    else
+    {
+        LOG_ERROR("Could not create user " + username);
+    }
+    return retCode;
+}
+
+bool User::deleteUser(const String& username)
+{
+    bool retCode = false;
+    int8_t userIdx = getUserIdx(username);
+
+    if (-1 != userIdx)
+    {
+        m_registeredUsers[userIdx] = nullptr;
+        m_numberOfRegisteredUsers--;
+        delete m_registeredUsers[userIdx];
+        retCode = true;
+    }
+    return retCode;
+}
+
+void User::serialize(String& serialized)
+{
+    /*
+    Reserve memory on heap for JSON structure.
+    The data is not copied into the DynamicJsonDocument by default.
+    The value specified in DOC_SIZE has been computed with the help of the ArduinoJson Assistant v6
+    which is accessible at: https://arduinojson.org/v6/assistant/
+    */
+    const size_t DOC_SIZE = 16384;
```

Comment:
Sounds high?

Reply:
It is. I calculated this size for a fully packed user database which includes 8 users with max username length of 16 chars, max 128 password/hash chars and 32 existing permissions per user